doc/designs/quic-design/dgram-api.md

```diff
@@ -29,6 +29,105 @@ subset of `struct mmsghdr` as being supported, specifically no control messages;
 The flags argument is defined by us. Initially we can support something like
 `MSG_DONTWAIT` (say, `BIO_DONTWAIT`).
 
+### Implementation Questions
+
+If we go with this, there are some issues that arise:
+
+- Are `BIO_mmsghdr`, `BIO_msghdr` and `BIO_iovec` simple typedefs
+  for OS-provided structures, or our own independent structure
+  definitions?
+
+  - If we use OS-provided structures:
+
+    - We would need to include the OS headers which provide these
+      structures in our public API headers.
+
+    - If we choose to support these functions when OS support is not available
+      (see discussion below), We would need to define our own structures in this
+      case (a “polyfill” approach).
+
+  - If we use our own structures:
+
+    - We would need to translate these structures during every call.
+
+      But we would need to have storage inside the BIO_dgram for *m* `struct
+      msghdr`, *m\*v* iovecs, etc. Since we want to support multithreaded use
+      these allocations probably will need to be on the stack, and therefore
+      must be limited.
+
+      Limiting *m* isn't a problem, because `sendmmsg` returns the number
+      of messages sent, so the existing semantics we are trying to match
+      lets us just send or receive fewer messages than we were asked to.
+
+      However, it does seem like we will need to limit *v*, the number of iovecs
+      per message. So what limit should we give to *v*, the number of iovecs? We
+      will need a fixed stack allocation of OS iovec structures and we can
+      allocate from this stack allocation as we iterate through the `BIO_msghdr`
+      we have been given. So in practice we could just only send messages
+      until we reach our iovec limit, and then return.
+
+      For example, suppose we allocate 64 iovecs internally:
+
+      ```c
+      struct iovec vecs[64];
+      ```
+
+      If the first message passed to a call to `BIO_writem` has 64 iovecs
+      attached to it, no further messages can be sent and `BIO_writem`
+      returns 1.
+
+      If three messages are sent, with 32, 32, and 1 iovecs respectively,
+      the first two messages are sent and `BIO_writem` returns 2.
+
+      So the only important thing we would need to document in this API
+      is the limit of iovecs on a single message; in other words, the
+      number of iovecs which must not be exceeded if a forward progress
+      guarantee is to be made. e.g. if we allocate 64 iovecs internally,
+      `BIO_writem` with a single message with 65 iovecs will never work
+      and this becomes part of the API contract.
+
+      Obviously these quantities of iovecs are unrealistically large.
```

Comment:
`s/large/small/`

Reply:
I could easily see more than 64 being required.  E.g. a database query that returns multiple results.  Writing each record as a separate iovec will avoid a copy/gather.  It could even be each field in each record.

Reply:
We're talking about UDP packets here. Anything sending over UDP is going to need to have some sort of chunking mechanism of its own to meet MTU. Nobody is going to be sending database results directly over a BIO_dgram.

The main use of iovec as I see it is for concatenating a single header buffer and the encrypted payload. Even a value of two would suffice for that application.

Reply:
I've seen some protocols (the one I remember most was from Microsoft) that use 4-6 iovecs for some datagram-protocol messages (roughly like "cleartext header, main payload, auxiliary payload, cryptographic tag"), but that basically is reinforcing Hugo's point.